serde/src/java/org/apache/hadoop/hive/serde2/objectinspector/primitive/JavaDateObjectInspector.java

```diff
@@ -38,7 +38,7 @@ public DateWritableV2 getPrimitiveWritableObject(Object o) {
 
   @Override
   public Date getPrimitiveJavaObject(Object o) {
-    return o == null ? null : (Date) o;
+    return o == null ? null : Date.valueOf(o.toString());
```

Comment:
If this is the correct fix, I believe it's also needed on the `new DateWritableV2((Date) o)` in the method above. I'm having code break there on a write path.